### PR TITLE
Add pagination and TOC Filter in quality control view

### LINKED ISSUE #1166: Entries QA: Fetch entries using pagination

### LINKED ISSUE #1167: Entries QA: Transform ToC filter to match with current entry matrix filters

### PR DESCRIPTION
**Pull request by samshara**
fixes #1166 
fixes #1167

styling required

### CODE REVIEW

**Comment by samshara** [comment]
@frozenhelium the entry edit does not work when used with data pulled from the server. This might be due to the inconsistency of lead information present in each entry which slightly differs from that of the original lead information.

File: src/components/general/EntryVerify/index.tsx [hunk 1]
```diff
@@ -117,16 +121,11 @@ function EntryVerify(props: Props) {
...
-    useEffect(() => {
-        setVerificationStatus(value);
-    }, [value]);
-
-    // FIXME: memoize this
-    const selectedOption = verificationStatusOptions.find(v => v.key === verified) ||
-                           verificationStatusOptions[1];
+    const selectedOption = useMemo(
+        () => verificationStatusOptions.find(v => v.key === value) ||
+            verificationStatusOptions[1],
```
Comment (tnagorra) [inline src/components/general/EntryVerify/index.tsx]:
```suggestion
        () => verificationStatusOptions.find(v => v.key === value) ??
            verificationStatusOptions[1],
```

Also why not use `0`?

File: src/views/Entries/QualityControl/EntryCard/index.tsx [hunk 2]
```diff
@@ -97,6 +96,12 @@ function EntryCard(props: EntryCardProps) {
         // TODO; disable all actions if pending
     }, []);
 
+    const [isVerified, setVerificationStatus] = React.useState<boolean>(entry.verified);
+
+    const handleEntryVerify = React.useCallback((status: boolean) => {
+        setVerificationStatus(status);
+    }, []);
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/EntryCard/index.tsx]:
```suggestion
    const handleEntryVerify = setVerificationStatus;
```

File: src/views/Entries/QualityControl/index.tsx [hunk 3]
```diff
@@ -78,29 +139,30 @@ function QualityControl(props: QualityControlProps) {
...
-                                    entry={e}
-                                    lead={lead}
-                                    framework={framework}
-                                    isDeleted={deletedEntries[e.id]}
-                                    onDelete={handleEntryDelete}
-                                />
-                            ));
-                        })}
+                        { pending && <LoadingAnimation /> }
+                        {response && response.results.map(e => (
```
Comment (AdityaKhatri) [inline src/views/Entries/QualityControl/index.tsx]: Use List component

File: src/components/TableOfContents/index.tsx [hunk 4]
```diff
@@ -7,12 +7,13 @@ import ListView from '#rsu/../v2/View/ListView';
 import styles from './styles.scss';
 
 interface Props<T, K extends string | number>{
-    keySelector: (datum: T) => K;
+    idSelector: (datum: T) => K;
+    keySelector: (datum: T) => K | undefined;
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx:13]: What is the difference between key and id?

File: src/views/Entries/QualityControl/index.tsx [hunk 5]
```diff
@@ -78,29 +139,30 @@ function QualityControl(props: QualityControlProps) {
...
+                                framework={framework}
+                                isDeleted={deletedEntries[e.id]}
+                                onDelete={handleEntryDelete}
+                            />
+                        ))}
                     </div>
                 )}
             />
+            <footer className={styles.footer}>
+                <Pager
```
Comment (AdityaKhatri) [inline src/views/Entries/QualityControl/index.tsx:237]: Lets use the same Pager for both pages but switch the data according to which page the user is in.
Also, for QC lets save the selected matrix header and active page / total entries count in redux.

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (tnagorra) [reply src/components/TableOfContents/index.tsx:13]: Do we even need key?

File: src/components/general/EntryVerify/index.tsx (hunk 1 above)
Comment (samshara) [reply src/components/general/EntryVerify/index.tsx]: 0 is for verified

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (samshara) [reply src/components/TableOfContents/index.tsx:13]: the server API requires it. I don't know why but it is as such right now.

File: src/components/general/EntryVerify/index.tsx (hunk 1 above)
Comment (tnagorra) [reply src/components/general/EntryVerify/index.tsx]: I think we should not hard-code this in this way. 
Thoughts @frozenhelium @AdityaKhatri ?

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (tnagorra) [reply src/components/TableOfContents/index.tsx:13]: Is key for client and id for server? Why do we need both?
Anyway, we need to document this through comments before we can merge this functionality.

Comment (AdityaKhatri) [reply src/components/TableOfContents/index.tsx:13]: Each widget option has id and key which is prevelant throughout code, so i think its fine.
Still, I'll quickly add the reason behind having 2 things.

File: src/components/TableOfContents/index.tsx [hunk 6]
```diff
@@ -29,17 +33,24 @@ function ToCItem<T, K extends string | number>(props: ToCItemProps<T, K>) {
...
     const children = childrenSelector(option);
 
     const handleClick = useCallback(
-        () => onChange(id),
-        [id, onChange],
+        () => {
+            if (isDefined(key)) {
+                return onChange({ key, id });
+            }
+            return undefined;
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: Why are we returning? onChange has no return value.

File: src/components/general/EntryVerify/index.tsx [hunk 7]
```diff
@@ -117,16 +121,11 @@ function EntryVerify(props: Props) {
...
-    useEffect(() => {
-        setVerificationStatus(value);
-    }, [value]);
-
-    // FIXME: memoize this
-    const selectedOption = verificationStatusOptions.find(v => v.key === verified) ||
-                           verificationStatusOptions[1];
+    const selectedOption = useMemo(
+        () => verificationStatusOptions.find(v => v.key === value) ??
+            verificationStatusOptions[1],
```
Comment (tnagorra) [inline src/components/general/EntryVerify/index.tsx]: We should think about a better way to identify a default value @thenav56 @frozenhelium @AdityaKhatri

Comment (tnagorra) [reply src/components/general/EntryVerify/index.tsx]: At least, it shouldn't break silently.

File: src/redux/selectors/siloDomainData/entries.js [hunk 8]
```diff
@@ -30,6 +30,21 @@ export const entriesViewActivePageSelector = createSelector(
     entriesView => entriesView.activePage || 1,
 );
 
+export const qualityControlViewActivePageSelector = createSelector(
+    entriesViewForProjectSelector,
+    entriesView => entriesView.qcViewActivePage || 1,
```
Comment (tnagorra) [inline src/redux/selectors/siloDomainData/entries.js]: use `??`

File: src/redux/selectors/siloDomainData/entries.js [hunk 9]
```diff
@@ -30,6 +30,21 @@ export const entriesViewActivePageSelector = createSelector(
...
+);
+
+export const qualityControlViewEntriesCountSelector = createSelector(
+    entriesViewForProjectSelector,
+    entriesView => entriesView.qcViewEntriesCount || 0,
```
Comment (tnagorra) [inline src/redux/selectors/siloDomainData/entries.js]: Use `??`

File: src/views/Entries/QualityControl/EntryCard/index.tsx [hunk 10]
```diff
@@ -91,12 +95,22 @@ function EntryCard(props: EntryCardProps) {
         isDeleted,
     } = props;
 
+    console.warn('here', lead);
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/EntryCard/index.tsx]: console.warn on render meant for debuggin.

File: src/views/Entries/QualityControl/EntryCard/index.tsx [hunk 11]
```diff
@@ -91,12 +95,22 @@ function EntryCard(props: EntryCardProps) {
...
+    const {
+        url: leadUrlFromProps,
+        attachment,
+    } = lead;
+
+    const leadUrl = (attachment && attachment.file) || leadUrlFromProps;
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/EntryCard/index.tsx]: Use `??`

File: src/views/Entries/QualityControl/index.tsx [hunk 12]
```diff
@@ -57,52 +113,132 @@ function QualityControl(props: QualityControlProps) {
...
+        body: {
+            filters: [
+                selected && Object.values(selected),
+                ...processedFilters,
+                ['project', projectId],
+            ].filter(isDefined),
+        },
+        method: 'POST',
+        onSuccess: (successResponse) => {
+            console.warn('here', successResponse);
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: console.warn meant for debugging

File: src/views/Entries/QualityControl/index.tsx [hunk 13]
```diff
@@ -57,52 +113,132 @@ function QualityControl(props: QualityControlProps) {
...
+            setEntriesCount({ count: successResponse.count });
+        },
+    });
+
+    useEffect(
+        () => getEntries(),
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]:
```suggestion
       getEntries,
```

File: src/views/Entries/QualityControl/index.tsx [hunk 14]
```diff
@@ -57,52 +113,132 @@ function QualityControl(props: QualityControlProps) {
...
+    const handleEntryDelete = React.useCallback((entryId) => {
+        setDeletedEntries(oldDeletedEntries => ({ ...oldDeletedEntries, [entryId]: true }));
+    }, [setDeletedEntries]);
+
+    const handleSelection = useCallback((value) => {
+        if (selected && selected.id === value.id) {
+            setSelection({ matrixKey: undefined });
+        } else {
+            setSelection({ matrixKey: value });
+        }
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: We can use a ternary operator here

File: src/components/TableOfContents/index.tsx (hunk 6 above)
Comment (samshara) [reply src/components/TableOfContents/index.tsx]: some of TOC elements don't have the corresponding key and aren't filterable. undefined is returned as to be consistent with eslint rules.

Comment (tnagorra) [reply src/components/TableOfContents/index.tsx]: The gist is that we don't need to return anything. Why return at all?

File: src/components/general/EntryVerify/index.tsx (hunk 1 above)
Comment (AdityaKhatri) [reply src/components/general/EntryVerify/index.tsx]: Not ideal, but works right now

File: src/typings/entry.ts [hunk 15]
```diff
@@ -67,7 +69,12 @@ export interface EntryFields extends DatabaseEntityBase {
     image?: string;
     tabularField: number;
     tabularFieldData: TabularDataFields;
-    lead: number;
+    lead: Pick<Lead, EntryLeadType> & {
+        assingee: number[];
```
Comment (tnagorra) [inline src/typings/entry.ts]:
```suggestion
        assignee: number[];
```

File: src/components/general/EntryVerify/index.tsx (hunk 1 above)
Comment (tnagorra) [reply src/components/general/EntryVerify/index.tsx]: "Works right now" doesn't work for me.

File: src/components/TableOfContents/index.tsx (hunk 4 above)
Comment (tnagorra) [reply src/components/TableOfContents/index.tsx:13]: Where was the reason added?

**Pull request merged**